File: programmers/level2/42583.js
// https://school.programmers.co.kr/learn/courses/30/lessons/42583
// 다리를 지나는 트럭
// 해답:https://velog.io/@chloeelog/%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%A8%B8%EC%8A%A4-%EB%8B%A4%EB%A6%AC%EB%A5%BC-%EC%A7%80%EB%82%98%EB%8A%94-%ED%8A%B8%EB%9F%AD-JavaScript


function solution(bridge_length, weight, truck_weights) {
  let time = 0;
  let bridge = Array(bridge_length).fill(0);
  let bridge_sum = 0;

  time++;
  bridge.shift();
  bridge_sum += truck_weights[0];
  bridge.push(truck_weights.shift());
  console.log('time', time);
  console.log('bridge', bridge);
  console.log('bridge_sum', bridge_sum);
  while (bridge_sum > 0) {
    time++;
    bridge_sum -= bridge.shift(); //[6]
    console.log('time', time);
    console.log('bridge', bridge);
    console.log('bridge_sum', bridge_sum);

    if (truck_weights.length > 0 && bridge_sum + truck_weights[0] <= weight) {
      bridge_sum += truck_weights[0];
      bridge.push(truck_weights.shift()); //[0,6]
      console.log('bridge_sum', bridge_sum);
    } else {
      bridge.push(0); //[6,0]
    }
    console.log('bridge', bridge);
  }

  return time;
}

console.log(solution(2, 10, [7, 4, 5, 6])); //8
// console.log(solution(100, 100, [10])); //101
// console.log(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10])); //110

File: programmers/python/level2/42583.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
  answer = 0
  ing_arr =deque([])
  t_len=len(truck_weights)
  t_idx=0
  while t_idx < t_len:
    # print('-'*50)
    # print('1 t_idx',t_idx)
    # print('ing_arr',ing_arr)
    # print('len(ing_arr)',len(ing_arr))

    while len(ing_arr) < bridge_length:
      answer+=1
      # print('answer',answer)
      # print('1 t_idx',t_idx)
      if t_idx < t_len and sum(ing_arr) + truck_weights[t_idx] <= weight:
        ing_arr.append(truck_weights[t_idx])
        t_idx+=1
      else:
        ing_arr.append(0)
      # print('ing_arr',ing_arr)

    if t_idx < t_len:
      out= ing_arr.popleft()
      # print('while out',out)

  # print('f ing_arr',ing_arr)
  while sum(ing_arr) > 0:
    out= ing_arr.popleft()
    # print('f out',out)
    answer+=1

  return answer

print(solution(2,	10,	[7,4,5,6]))
# print(solution(100,	100,	[10]))
# print(solution(100,	100,	[10,10,10,10,10, 10,10,10,10,10]))
# print(solution(100,	100,	[1,2,3,4,5,6,7,8,9,10]))

# 2	10	[7,4,5,6]	8
# 100	100	[10]	101
# 100	100	[10,10,10,10,10,10,10,10,10,10]	110
